### DISCUSSION TITLE
Obtain only visible lines of mesh.extract_feature_edges()

### DISCUSSION THREAD

**Discussion by thnake** [pyvista/pyvista#3040] [Q&A] [answered]
Hello,

I would like to extract the subset of indices that describe the visible feature edges, i.e. a subset of the array  `edges = mesh.extract_feature_edges()` . For example, in the images below, I want the to get the coordinates of indices of the red lines.

![](https://user-images.githubusercontent.com/5691313/180266074-fb8cef4c-5090-4db3-a8f8-0986684108fa.png )

![grafik](https://user-images.githubusercontent.com/5691313/180266305-5171cbd0-91b4-4b8f-a317-fc02bcef12d1.png )

Thanks!

**Comment by akaszynski** [comment]
This is a great question, and I think you've just highlighted a limitation of both PyVista and VTK.

If you wanted the cell or point indices that were attached to the edges we can just add them as cell data and then report them in a plot:

```py
import pyvista as pv
from pyvista import examples
pv.set_plot_theme('document')

mesh = examples.download_cow()
mesh.cell_data['cell_ids'] = range(mesh.n_cells)

edges = mesh.extract_feature_edges()
edges.plot(line_width=5)
```

![tmp](https://user-images.githubusercontent.com/11981631/180331481-665c2470-2a97-463e-ac10-e0e536d203af.png)


However, I can't attach data directly to lines. Actually, with the way this mesh is being represented, there's actually no lines (edges really) in the first place. It's all triangles:

```py
>>> mesh.faces
array([  4, 250, 251, ..., 966, 961, 970])
>>> mesh.lines
array([], dtype=int64)
```

The extracted edges, are as you expect, lines:
```py
>>> edges.lines
array([   2,    0,    1, ...,    2, 1108,  391])
```

---
So to answer your question:

> I would like to extract the subset of indices that describe the visible feature edges

You can extract either the cell or point indices (just use `point` instead of `cell` in the example above), but you can't extract the line indices since the original mesh doesn't even contain lines. Even if you added them in, VTK still wouldn't extract them and even if we could we associate any data with them.

Hope that (mostly) helps.

**Reply by thnake** [reply]
Hi, is there a way to extract only the cells that are rendered on the screen, i.e. those that are not occluded? Maybe I could call `extract_feature_edges` on those

**Comment by thnake** [comment, marked as answer]
based on some code snipped from this site (I lost the link), I came up with this:

```python
plotter.add_mesh(mesh)

edges = mesh.extract_feature_edges()


selector = vtk.vtkOpenGLHardwareSelector()
selector.SetFieldAssociation(vtk.vtkDataObject.FIELD_ASSOCIATION_CELLS)
selector.SetRenderer(plotter.renderer)

sz = plotter.size()
selector.SetArea(0,0,sz.width(),sz.height())
selection = selector.Select()
extractor = vtk.vtkExtractSelection()

extractor.SetInputData(0, mesh)
extractor.SetInputData(1, selection)
extractor.Update()

visible = pv.wrap(extractor.GetOutput())
visibleEdges = visible.extract_feature_edges()
# plotter.clear()
plotter.add_mesh(visibleEdges, color="red", line_width=5)
```

**Reply by akaszynski** [reply]
I totally missed your”visible” remark. I see what you’re going for now. Thanks for posting your solution!

**Comment by jh27kim** [comment]
Hi, 
I also have been looking for this. 
When I try to run the code snippet below, it raises two errors of which

i) ```sz = plotter.size() ```
"'Plotter' object has no attribute 'size'"
When I plug in any arbitrary integer and proceed, there is another error

ii) ```selection = selector.Select() ```
"Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)"

```
import pyvista as pv
from pyvista import examples
import vtk

plotter = pv.Plotter()

mesh = examples.download_cow()

plotter.add_mesh(mesh)

edges = mesh.extract_feature_edges()

selector = vtk.vtkOpenGLHardwareSelector()
selector.SetFieldAssociation(vtk.vtkDataObject.FIELD_ASSOCIATION_CELLS)
selector.SetRenderer(plotter.renderer)

# sz = plotter.size() # ERROR:
# selector.SetArea(0,0,sz.width(),sz.height())
selector.SetArea(0,0,1024,1024) # plotter has no attribute as size()

# ERROR:Segmentation Error SIGSEGV
# Exit code 139
selection = selector.Select()
extractor = vtk.vtkExtractSelection()

extractor.SetInputData(0, mesh)
extractor.SetInputData(1, selection)
extractor.Update()

visible = pv.wrap(extractor.GetOutput())
visibleEdges = visible.extract_feature_edges()
# plotter.clear()
plotter.add_mesh(visibleEdges, color="red", line_width=5)
```


Versions I am currently using are:
python                   3.9.16
pyvista                   0.38.5
vtk                          9.2.6

Appreciate any feedback
